sync/transactions.go

```diff
@@ -179,70 +156,179 @@ func makeResponse(now time.Time, diff int64, trans int32, lastSeq int32) *Report
-	transactions[k] = t
+	transactions[getTransactionKey(t.Addr.IP, t.ID)] = t
+	tMutex.Unlock()
 }
 
-func recordNewRequest(req *ReportRequest, now time.Time, srcAddr *net.UDPAddr) {
-	startTransaction(req.Trans, srcAddr, now, req.Name)
-	//now create the TripTimes array to include the estimated number of triptimes
+func (t *Transaction) Untrack() {
 	tMutex.Lock()
-	t, _ := transactions[getTransactionKey(srcAddr.IP, req.Trans)]
-	defer tMutex.Unlock()
-	t.TripTimes = make([]time.Duration, 0, config.Iterations+1)
-	t.Offsets = make([]int64, 0, config.Iterations+1)
-	//store the first offset and we'll calculate the trip time when we get the first response
-	diff := req.Time - now.UnixNano()
-	t.Offsets = append(t.Offsets, diff)
+	delete(transactions, getTransactionKey(t.Addr.IP, t.ID))
+	tMutex.Unlock()
 }
 
-func correctReply(replyAddr string, srcAddr *net.UDPAddr) (*net.UDPAddr, error) {
-	//determine the addr to send based on the received message
-	destAddr, err := net.ResolveUDPAddr("udp", replyAddr)
-	if err != nil {
-		return nil, err
+func (t *Transaction) NewJob(j *Job) {
+	t.jobCh <- j
+}
+
+func (t *Transaction) SeqMatches(seq int32) bool {
+	return t.seq == seq
+}
+
+func (t *Transaction) ReaderSpin() {
+	var n int
+	var src *net.UDPAddr
+	var err error
+	var j *Job
+	buf := make([]byte, BufferSize)
+	for {
+		t.conn.SetReadDeadline(time.Now().Add(time.Second * 10))
+		n, src, err = t.conn.ReadFromUDP(buf)
+		if err != nil {
+			if nerr, _ := err.(net.Error); nerr.Temporary() || nerr.Timeout() {
+				if nerr.Timeout() {
+					//if we timed out then master/slave must've died or we're done
+					break
+				} else {
+					//its "temporary" so continue
+					continue
+				}
+			}
+			llog.Warn("error reading from UDP connection", llog.KV{"err": err})
+			break
+		}
+
+		j, err = DecodeMessage(buf[0:n], src)
+		if err != nil {
+			llog.Warn("error calling DecodeMessage in ReaderSpin", llog.KV{"err": err, "src": src})
+			break
+		}
+		go t.NewJob(j)
 	}
-	//if they're listening on all interfaces then use the srcAddr
-	if destAddr.IP.IsUnspecified() {
-		destAddr.IP = srcAddr.IP
+	t.Close()
+}
+
+func (t *Transaction) JobSpin() {
+	var j *Job
+	var p []byte
+	var err error
+	for j = range t.jobCh {
+		if !t.Addr.IP.Equal(j.SrcAddr.IP) {
+			llog.Warn("received job from a mismatched IP", llog.KV{
+				"src":      j.SrcAddr,
+				"expected": t.Addr,
+			})
+			continue
+		}
+
+		//we have to wrap this particular set in a mutex since removeOldTransactions reads this
+		tMutex.Lock()
+		t.LastResponse = time.Now()
+		tMutex.Unlock()
+
+		if j.MType == Init {
+			p, err = getReportBytes(makeReport(time.Now()))
+			if err != nil {
+				llog.Error("error from GetReportBytes", llog.KV{"err": err})
+				break
+			}
+		} else {
+			//increase seq for read
+			t.seq++
+			p = handleJob(j, t)
+			if p == nil {
+				//nothing left to write so this transaction is done
+				break
+			}
+		}
+		t.write(p)
 	}
-	return destAddr, nil
+	t.Close()
```

Comment:
if we're at this point it means `jobCh` is closed, meaning something else must have called `Close()` already right?